File: JavaScript/3_longest_substring_without_repeating_characters/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var start = 0,
        maxLen = 0,
        dict = {};
    
    for (var i=0; i<s.length; i++) {
        dict[s[i]] = -1;
    }   
    for (var i=0; i<s.length; i++) {
        if (dict[s[i]] !== -1) {
            while (start <= dict[s[i]]) {
                dict[s[start]] = -1;
                start += 1;
            }
        }
        if (i-start+1 > maxLen) {
           maxLen = i - start + 1;  
        }
        dict[s[i]] = i;
    } 
    return maxLen;
};

File: Cpp/3_longest_substring_without_repeating_characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int result = 0;
        if (s.size() == 0) return result;
        
        int start = 0;
        map<int, int> m;
        for (auto it = s.begin(); it != s.end(); it ++) m[*it] = -1;
        for (int i = 0; i < s.size(); i ++)
        {
            if (m[s[i]] != -1)
            {
                while (start <= m[s[i]]) m[s[start ++]] = -1;
            }
            result = (result > i-start+1) ? result : i-start+1;
            m[s[i]] = i;
        }
        return result;
    }
};